src/astlowering/Uncurry.ml

```diff
@@ -125,9 +439,167 @@ module ScillaCG_Uncurry = struct
+              uncurry_func_typ tp (List.length params'));
+      }
+    in
+    let%bind fun_rep =
+      match fst fe with
+      | GasExpr (_, (_, rep)) -> pure @@ translate_eannot rep
+      | _ ->
+          fail0
+            "Uncurry: internal error: uncurry_func_epxr has no Gas rapped \
+             around Fun."
+    in
+    let fun_rep' =
+      {
+        fun_rep with
+        ea_tp =
+          Option.map fun_rep.ea_tp ~f:(fun tp ->
+              uncurry_func_typ tp (List.length params'));
+      }
+    in
+    pure
+      ( UCS.GasExpr (StaticCost static_gas, (UCS.Fun (params', body'), fun_rep')),
+        new_rep' )
+
+  (* Note:
+     * Functions are added to the env when they are uncurried
+     * Functions are only removed from the environment when the variable name is redefined to
+     something else
+     * translate_expr does not return an environment because the scope of
+     newly defined functions in expr ends there
+  *)
+  (* If functions that are NOT uncurried, their App will be transformed apply arguments
+     one by one *)
+  (* Note: ienv contains variables in UCS syntax *)
+  and translate_expr newname (e, annot) ienv =
+    let rec go_expr (e, erep) ienv =
       match e with
+      | Let (i, topt, lhs, rhs) ->
+          (* If the function is to be uncurried *)
+          if Option.is_some (Identifier.get_rep i).ea_auxi then
+            let%bind uncurried_lhs = uncurry_func_epxr newname ienv lhs in
+            let uncurried_ty = (snd uncurried_lhs).ea_tp in
+            (* Translate i to have the new uncurried type in rep *)
+            let translated_i =
+              let rep = translate_eannot (Identifier.get_rep i) in
+              let rep' = { rep with ea_tp = uncurried_ty } in
+              Identifier.mk_id (Identifier.get_id i) rep'
+            in
+            (* Add i into ienv to run rhs *)
+            let%bind rhs' =
+              go_expr rhs ((translated_i, uncurried_ty) :: ienv)
+            in
+            let topt' =
+              match topt with None -> None | Some _ -> uncurried_ty
+            in
+            pure
+              ( UCS.Let (translated_i, topt', uncurried_lhs, rhs'),
+                translate_eannot erep )
+          else
+            let translated_i = translate_var i in
+            (* We don't uncurry it - and we make sure it's not in the ienv *)
+            let ienv' =
+              List.filter ienv ~f:(fun (iden, _) ->
+                  not @@ Identifier.equal translated_i iden)
+            in
+            let%bind lhs' = go_expr lhs ienv in
+            let%bind rhs' = go_expr rhs ienv' in
+            pure
+              ( UCS.Let
+                  (translated_i, Option.map ~f:translate_typ topt, lhs', rhs'),
+                translate_eannot erep )
+      | App (a, l) ->
+          let a' = translate_var a in
+          (* If the function was uncurried *)
+          if Identifier.is_mem_id a' (fst @@ List.unzip ienv) then
```

Comment:
This involves two traversals over the list. It's simpler to just search for `'a` in the zipped list by using `List.mem` directly